service/storage_service.cc

```diff
@@ -1809,8 +1809,13 @@ future<> storage_service::join_topology(sharded<db::system_distributed_keyspace>
         }
     }
 
-    // if the node is bootstrapped the function will do nothing since we already created group0 in main.cc
-    ::shared_ptr<group0_handshaker> handshaker = raft_topology_change_enabled()
+    // setup_group0 will do nothing if the node has already set up group 0 in setup_group0_if_exist in main.cc,
+    // which is the case when the node is restaring and not joining the new group 0 in Raft recovery.
+    //
+    // We use the legacy handshaker in Raft recovery to join the new group 0 without involving the topology
+    // coordinator.
+    ::shared_ptr<group0_handshaker> handshaker =
+            raft_topology_change_enabled() && _db.local().get_config().recovery_leader().empty()
```

Comment:
so if we leave `recovery_leader` specified, we'll create the wrong handshaker here after recovery is finished, but that's not a problem, because the handshaker won't be used --- which is what the previous paragraph tries to explain?

Perhaps the order of paragraphs should be reversed, you could also make it more explicit that creating the wrong handshaker is not a problem

Reply:
> so if we leave `recovery_leader` specified, we'll create the wrong handshaker here after recovery is finished, but that's not a problem, because the handshaker won't be used --- which is what the previous paragraph tries to explain?

Yes, but the previous paragraph tries to explain that creating a wrong handshaker doesn't matter is not a problem in every case when group 0 is already set up, not only in the new recovery procedure.

> Perhaps the order of paragraphs should be reversed, you could also make it more explicit that creating the wrong handshaker is not a problem

Not sure about the order since the first paragraph is more general, but I'll try to make this comment more clear somehow.